sonic-net/sonic-buildimage
# Added BUILD flags to provision for building the kernel with non-upstream patches

Signed-off-by: Vivek Reddy <vkarri@nvidia.com>



#### Why I did it

Provide a way to include non-upstream kernel patches in the kernel build process

Requires https://github.com/sonic-net/sonic-linux-kernel/pull/296 and can be merged in any order

#### How I did it

Introduced two new environment variables:
```
EXTERNAL_KERNEL_PATCH_URL = <url to the archive containing patches>, empty by default
INCLUDE_EXTERNAL_PATCH_TAR= <y|n to include the patches under platform/ folder> 'n' by default
```

Note: For `INCLUDE_EXTERNAL_PATCH_TAR` to work, the platform has to override `EXTERNAL_KERNEL_PATCH_TAR `variable. And the variable should be a path pointing to the tarball containing patches

Use SONIC_OVERRIDE_BUILD_VARS to fuse in these env variables:

```
- NOJESSIE=1 NOSTRETCH=1 NOBUSTER=1 make SONIC_OVERRIDE_BUILD_VARS=' INCLUDE_EXTERNAL_PATCH_TAR=y ' target/debs/bullseye/linux-headers-5.10.0-12-2-common_5.10.103-1_all.deb
- NOJESSIE=1 NOSTRETCH=1 NOBUSTER=1 make SONIC_OVERRIDE_BUILD_VARS=' EXTERNAL_KERNEL_PATCH_URL=http://<url-root>/loc/patches.tar.gz ' target/debs/bullseye/linux-headers-5.10.0-12-2-common_5.10.103-1_all.deb
```

#### Structure of the patch tar archive
1) It should contain a file named series. series should provide an order of which the patches have to be applied
2) All the patches should be present in the archive in the same folder where series resides.
3) Developers should make sure patches apply cleanly over the existing patches present in the sonic-linux-kernel/ repo. 
```
admin@build-server:/sonic-buildimage/platform/mellanox/non-upstream-patches$ tar -tf patches.tar.gz
./
./mlx5-Implement-get_module_eeprom_by_page.patch
./mlx5-Add-support-for-DSFP-module-EEPROM-dumps.patch
./series
./mlx5-Refactor-module-EEPROM-query.patch
```

#### How to verify it

1) Build the kernel
2) Check if the patches are applied on the kernel source
```
vkarri@build-server:/sonic-buildimage/src/sonic-linux-kernel/linux-5.10.103$ git log --oneline -n 4
a524105fc (HEAD -> master, refs/patches/master/mlx5-Add-support-for-DSFP-module-EEPROM-dumps.patch) From 4c88fa412a100f925b8ab1aa952a672895f69d35 Mon Sep 17 00:00:00 2001 Subject: net/mlx5: Add support for DSFP module EEPROM dumps

8f366b27c (refs/patches/master/mlx5-Implement-get_module_eeprom_by_page.patch) From e109d2b204daa223e6d3cdaa369071c3ea96dcbf Mon Sep 17 00:00:00 2001 Subject: net/mlx5: Implement get_module_eeprom_by_page()

eb9bddfad (refs/patches/master/mlx5-Refactor-module-EEPROM-query.patch) From e19b0a3474ab9ef90dd110af9f39fc87329755f1 Mon Sep 17 00:00:00 2001 Subject: net/mlx5: Refactor module EEPROM query

cf79fff3a (refs/patches/master/armhf_secondary_boot_online.patch) From 500d48948f43d18a07894b5fd4f891afbade1d79 Mon Sep 17 00:00:00 2001
```


#### Which release branch to backport (provide reason below if selected)




#### Description for the changelog


#### Ensure to add label/tag for the feature raised. example - PR#2174 under sonic-utilities repo. where, Generic Config and Update feature has been labelled as GCU.

#### Link to config_db schema for YANG module changes


#### A picture of a cute animal (not mandatory but encouraged)
